relay/nakama/siwe/siwe.go

```diff
@@ -0,0 +1,276 @@
+type GenerateResult struct {
+	SIWEMessage string `json:"siwe_message"`
+}
+
+// nonceStorageObj is the struct that is serizlied to JSON and saved to the Nakama storage layer. Each signer address
+// maps to a different nonceStorageObj. A signature is valid if the nonce contained in the signed message does NOT
+// already have an entry in this object.
+type nonceStorageObj struct {
+	Nonces  []*nonceWindow `json:"nonces"`
+	Version string         `json:"-"`
+}
+
+// nonceWindow represents a single nonce that was used as well as the window of time that it was valid for.
+type nonceWindow struct {
+	Nonce     string    `json:"nonce"`
+	IssuedAt  time.Time `json:"issued_at"`
+	ExpiredAt time.Time `json:"expired_at"`
+}
+
+// ValidateSignature ensures the given message is valid and the given signature is actually a signature for the message.
+func ValidateSignature(ctx context.Context, nk runtime.NakamaModule, signerAddress, message, signature string) error {
+	msg, err := siwe.ParseMessage(message)
+	if err != nil {
+		return eris.Wrap(err, "failed to parse message")
+	}
+
+	if msg.GetDomain() != DefaultDomain {
+		return eris.New("domain is incorrect")
+	}
+	if uri := msg.GetURI(); uri.String() != DefaultURI {
+		return eris.New("uri is incorrect")
+	}
+
+	nonce, err := getNonceWindow(msg)
+	if err != nil {
+		return eris.Wrap(err, "nonce does not have a valid time window")
+	}
+
+	// Make sure the signature is valid.
+	pubKey, err := msg.VerifyEIP191(signature)
+	if err != nil {
+		return eris.Wrap(err, "failed to verify signature")
+	}
+
+	// Make sure the signer of the message and the signer in the message match
+	actualSignerAddress := crypto.PubkeyToAddress(*pubKey).Hex()
+	if actualSignerAddress != signerAddress {
+		return eris.Wrap(err, "signer address in message and signature are different ")
+	}
+
+	// Make sure the nonce is valid and hasn't yet been used.
+	nonceUsed, err := useNonce(ctx, nk, actualSignerAddress, nonce)
+	if err != nil {
+		return eris.Wrap(err, "failed to use nonce")
+	} else if !nonceUsed {
+		return eris.New("nonce has already been used")
+	}
+	return nil
+}
+
+func getNonceWindow(msg *siwe.Message) (*nonceWindow, error) {
+	// Make sure the expiration time is in the future
+	expString := msg.GetExpirationTime()
+	if expString == nil {
+		return nil, eris.New("missing expiration time")
+	}
+	expTime, err := time.Parse(time.RFC3339, *expString)
+	if err != nil {
+		return nil, eris.New("unable to parse expiration time")
+	}
+	if time.Now().After(expTime) {
+		return nil, eris.New("message has expired")
+	}
+
+	// Make sure the issued time is in the past
+	issuedAt, err := time.Parse(time.RFC3339, msg.GetIssuedAt())
+	if err != nil {
+		return nil, eris.New("failed to parse issued at time")
+	}
+	if time.Now().Before(issuedAt) {
+		return nil, eris.New("message must be issued in the past")
+	}
+
+	delta := expTime.Sub(issuedAt)
+	if delta > DefaultTimeout {
```

Comment:
why do we need this when we already have expiration time?

Reply:
This check means we're only going to accept SIWE messages that are valid for at most a 5 minute window. The messages we generate conform to this requirement, but there's nothing preventing a user from constructing their own SIWE Message with a much longer valid window. I want to avoid a situation where a user makes a SIWE Message that is valid from time 0 to the heat death of the universe.